pcf/particle/aws/sns/sns_topic.py

```diff
@@ -0,0 +1,213 @@
+# Copyright 2018 Capital One Services, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pcf.core.aws_resource import AWSResource
+from pcf.core import State
+from pcf.util import pcf_util
+from botocore.exceptions import ClientError
+
+
+class SNSTopic(AWSResource):
+
+    """
+    This is the implementation of Amazon's SNS.
+    """
+    flavor = "sns"
+
+    START_PARAM_FILTER = {
+        "Name",  # Required
+    }
+
+    DEFINITION_FILTER = {
+        "Name",
+        "Attributes",
+    }
+
+    START_ATTR = {
+        "DeliveryPolicy",
+        "DisplayName",
+        "Policy"
+    }
+
+    START_SUB = {
+        "Protocol",
+        "Endpoint",
+        "Attributes",
+        "ReturnSubscriptionArn"
+    }
+
+    equivalent_states = {
+        State.running: 1,
+        State.stopped: 0,
+        State.terminated: 0,
+    }
+
+    UNIQUE_KEYS = ["aws_resource.Name"]
+
+    def __init__(self, particle_definition, session=None):
+        super(SNSTopic, self).__init__(particle_definition=particle_definition, resource_name="sns", session=session)
+        self.topic_name = self.desired_state_definition.get("Name")
+        self._set_unique_keys()
+
+    def _set_unique_keys(self):
+        """
+        Logic that sets keys from state definition that are used to uniquely identify the sns topic
+
+        """
+        self.unique_keys = SNSTopic.UNIQUE_KEYS
+
+    def _start(self):
+        """
+        Creates the SNS Topic according to the desired definition. The attribute parameter is not supported in the
+        create_topic() method in boto3 v1.9.5 and lower. Desired attributes are put in the custom_config dictionary
+        and added to the topic as a separate call, set_topic_attributes().
+
+        Returns:
+            response of boto3(v1.9.5) create_topic
+        """
+        start_definition = pcf_util.param_filter(self.get_desired_state_definition(), SNSTopic.START_PARAM_FILTER)
+        response = self.client.create_topic(**start_definition)
+        self._arn = response.get("TopicArn")
+
+        # add attributes if applicable
+        if self.custom_config.get("Attributes"):
```

Comment:
u can pass in attributes to the create_topic so it shouldnt be in custom_config and instead in the regular definition

Reply:
moto doesn't support that

Reply:
if moto doesnt support it then you should use placebo to test then. you shouldnt change the particle for the tests.